core/01_core.js

```diff
@@ -221,59 +257,346 @@
+            setPromise(id),
+            unwrapOpError(eventLoopTick),
+          );
+          promise = handleOpCallTracing(opName, id, promise);
+          promise[promiseIdSymbol] = id;
+          return promise;
+        };
+        break;
+
+      case 7:
+        fn = function async_op_7(a, b, c, d, e, f, g) {
+          const id = nextPromiseId++;
+          try {
+            const maybeResult = originalOp(id, a, b, c, d, e, f, g);
+            if (maybeResult !== undefined) {
+              movePromise(id);
+              return unwrapOpResultNewPromise(id, maybeResult, async_op_7);
+            }
+          } catch (err) {
+            ErrorCaptureStackTrace(err, async_op_7);
+            return PromiseReject(err);
+          }
+          let promise = PromisePrototypeThen(
+            setPromise(id),
+            unwrapOpError(eventLoopTick),
+          );
+          promise = handleOpCallTracing(opName, id, promise);
+          promise[promiseIdSymbol] = id;
+          return promise;
+        };
+        break;
+
+      case 8:
+        fn = function async_op_8(a, b, c, d, e, f, g, h) {
+          const id = nextPromiseId++;
+          try {
+            const maybeResult = originalOp(id, a, b, c, d, e, f, g, h);
+            if (maybeResult !== undefined) {
+              movePromise(id);
+              return unwrapOpResultNewPromise(id, maybeResult, async_op_8);
+            }
+          } catch (err) {
+            ErrorCaptureStackTrace(err, async_op_8);
+            return PromiseReject(err);
+          }
+          let promise = PromisePrototypeThen(
+            setPromise(id),
+            unwrapOpError(eventLoopTick),
+          );
+          promise = handleOpCallTracing(opName, id, promise);
+          promise[promiseIdSymbol] = id;
+          return promise;
+        };
+        break;
+
+      case 9:
+        fn = function async_op_9(a, b, c, d, e, f, g, h, i) {
+          const id = nextPromiseId++;
+          try {
+            const maybeResult = originalOp(id, a, b, c, d, e, f, g, h, i);
+            if (maybeResult !== undefined) {
+              movePromise(id);
+              return unwrapOpResultNewPromise(id, maybeResult, async_op_9);
+            }
+          } catch (err) {
+            ErrorCaptureStackTrace(err, async_op_9);
+            return PromiseReject(err);
+          }
+          let promise = PromisePrototypeThen(
+            setPromise(id),
+            unwrapOpError(eventLoopTick),
+          );
+          promise = handleOpCallTracing(opName, id, promise);
+          promise[promiseIdSymbol] = id;
+          return promise;
+        };
+        break;
+
+      default:
+        throw new Error(
+          `Too many arguments for async op codegen (length of ${opName} was ${
+            originalOp.length - 1
+          })`,
+        );
+    }
+    return (ops[opName] = fn);
   }
 
   function opAsync2(name, arg0, arg1) {
     const id = nextPromiseId++;
```

Comment:
If we could get rid of the requirement that promises resolving immediately have no promise ID, we can skip the `movePromise` step. Unfortunately there's a lot of code requiring promise IDs and I'm not sure we can do this yet.

Reply:
I think we can pass promiseId to Rust but create the actual promise only after dispatching we didn't receive immediate response. WDYT?

Reply:
I see that you already did that. I think this is already an improvement that you are lazily creating promises only when they are needed and passing numbers shouldn't be that bad in terms of performance.